apps/design/design.md

```diff
@@ -0,0 +1,159 @@
+# MSAL Go Design Guide
+
+Author: John Doak(jdoak@microsoft.com)
+Contributors: 
+- Keegan Caruso(Keegan.Caruso@microsoft.com)
+- Joel Hendrix(jhendrix@microsoft.com)
+- Santiago Gonzalez(Santiago.Gonzalez@microsoft.com)
+
+## History
+
+The original code submitted for Go MSAL was a translation of either Java or .Net code.  This was done as a best effort by an intern who was attempting their first crack at Go.  It had a
+very interesting structure that didn't fit into Go style and made it difficult to understand or
+change. It used global locks, global variables, base type classes (mimicing inheritence), ...
+
+This probably should have be re-written from scratch, but we decided to try and do it in pieces.
+The lesson to be learned from this is that this type of refactor leads to re-writing the code 7 or 8 times instead of once. 
```

Comment:
anything that you feel is still not go idiomatic which you would have done differently had you started with an empty repo?

MSAL has certain api and quirks, where consistency is valued higher than best possible across libraries.

Reply:
Internal naming, some unmarshaling for things like time, etc....

Externally, it is now pretty close to idiomatic

Reply:
Also, to be fair, I'm still not "in love".  There is still a certain amount of clunkyness.  I'm not sure if what I dislike is how MSAL does things, that I want more clients per auth service layers, or internal weirdness...

The most important thing is that now, with the exception of like two type aliases, our internal and external are separated.  If we like the external surface, then we now can really change internals completely without harming the external surface.